lib/phoenix_live_view/html_engine.ex

```diff
@@ -443,12 +467,54 @@ defmodule Phoenix.LiveView.HTMLEngine do
     end
   end
 
-  defp pop_args({wrapper, meta, kw}) do
-    {let, kw} = Keyword.pop(kw, :let, quote do: _assigns)
-    {let, {wrapper, meta, kw}}
+  @doc false
+  def __unmatched_let__!(pattern, value) do
+    message = """
+    cannot match arguments sent from `render_block/2` against the pattern in `let`.
+
+    Expected a value matching `#{pattern}`, got: `#{inspect(value)}`.
+    """
+
+    stacktrace =
+      self()
+      |> Process.info(:current_stacktrace)
+      |> elem(1)
+      |> Enum.drop(2)
+
+    reraise(message, stacktrace)
+  end
+
+  defp raise_if_let!(let, file) do
+    with {_pattern, %{line: line}} <- let do
+      message = "cannot use `let` on a component without inner content"
+      raise CompileError, line: line, file: file, description: message
+    end
   end
 
-  defp pop_args(assigns) do
-    assigns
+  defp build_component_clauses(let, state) do
+    case let do
+      {pattern, %{line: line}} ->
+        quote line: line do
+          unquote(pattern) ->
+            unquote(invoke_subengine(state, :handle_end, []))
+        end
+        ++
+        quote line: line, generated: true do
+          other ->
+            message = """
+            cannot match arguments sent from `render_block/2` against the pattern in `let`.
+
+            Expected a value matching `#{unquote(Macro.to_string(pattern))}`, got: `#{inspect(other)}`.
+            """
+            raise RuntimeError, message
+
+            # Phoenix.LiveView.HTMLEngine.__unmatched_let__!(unquote(Macro.to_string(pattern)), other)
```

Comment:
We should reduce the amount of generated code, so calling the unmatched let function should be better. :)

Reply:
Oh, sure :)